src/imp/propagator_envoy.js

```diff
@@ -0,0 +1,187 @@
+
+const BINARY_PROTO = {
+    nested : {
+        lightstep : {
+            options : { go_package : 'lightsteppb' },
+            nested  : {
+                BinaryCarrier : {
+                    fields : {
+                        deprecated_text_ctx : {
+                            rule : 'repeated',
+                            type : 'bytes',
+                            id   : 1,
+                        },
+                        basic_ctx : { type : 'BasicTracerCarrier', id : 2 },
+                    },
+                },
+                BasicTracerCarrier : {
+                    fields : {
+                        trace_id      : { type : 'fixed64', id : 1 },
+                        span_id       : { type : 'fixed64', id : 2 },
+                        sampled       : { type : 'bool', id : 3 },
+                        baggage_items : { keyType : 'string', type : 'string', id : 4 },
+                    },
+                },
+            },
+        },
+    },
+};
+
+
+export default class EnvoyPropagator extends LightStepPropagator {
+    constructor(tracer) {
+        super(tracer);
+        this._tracer = tracer;
+        this._envoyHeaderKey = CARRIER_ENVOY_HEADER_KEY;
+        this._carrierPb = pb.Root.fromJSON(BINARY_PROTO);
+    }
+
+    inject(spanContext, carrier) {
+        if (!carrier) {
+            this._tracer._error('Unexpected null carrier in call to inject');
+            return;
+        }
+        if (typeof carrier !== 'object') {
+            this._tracer._error(
+                `Unexpected '${typeof carrier}' FORMAT_BINARY carrier in call to inject`,
+            );
+            return;
+        }
+
+        let basicContext = {
+            span_id       : long.fromString(spanContext._guid, 16),
+            trace_id      : long.fromString(spanContext._traceGUID, 16),
+            sampled       : true,
+            baggage_items : {},
+        };
+        spanContext.forEachBaggageItem((key, value) => {
+            basicContext.baggage_items[key] = value;
+        });
+
+        let binaryCarrier = this._carrierPb.lookupType('BinaryCarrier');
+
+        let payload = {
+            basic_ctx : basicContext,
+        };
+
+        let err = binaryCarrier.verify(payload);
+        if (err) {
+            this._tracer._error(`Invalid Span Context: ${err}`);
+            return null;
+        }
+        let msg = binaryCarrier.create(payload);
+        let buffer = binaryCarrier.encode(msg).finish();
+        let bufferString = pb.util.base64.encode(buffer, 0, buffer.length);
+        carrier[this._envoyHeaderKey] = bufferString;
+
+        return carrier;
+    }
+
+    extract(carrier) {
+        // Iterate over the contents of the carrier and set the properties
+        // accordingly.
+        let foundFields = 0;
+        let spanGUID = null;
+        let traceGUID = null;
+        let sampled = true;
+
+        if (carrier[this._envoyHeaderKey] === undefined) {
```

Comment:
I believe the same comment for inject about the carrier being a byte array applies here as well.